File: app/src/main/java/amagi82/modularcharactersheetcreator/ui/_extras/widgets/FocusAwareEditText.java
package amagi82.modularcharactersheetcreator.ui._extras.widgets;

import android.content.Context;
import android.util.AttributeSet;
import android.view.KeyEvent;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;
import android.widget.TextView;

import amagi82.modularcharactersheetcreator.ui._extras.widgets.callbacks.EditTextListener;
import amagi82.modularcharactersheetcreator.ui._extras.widgets.callbacks.SimpleTextWatcher;

public class FocusAwareEditText extends EditText {
    private EditTextListener listener;
    private String text;

    public FocusAwareEditText(Context context) {
        super(context);
        init();
    }

    public FocusAwareEditText(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public FocusAwareEditText(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    private void init() {
        setOnFocusChangeListener(new OnFocusChangeListener() {
            @Override
            public void onFocusChange(View view, boolean hasFocus) {
                if (hasFocus) listener.onKeyboardShown();
                else {
                    //Not focused. Hide keyboard and notify
                    InputMethodManager imm =  (InputMethodManager) getContext().getSystemService(Context.INPUT_METHOD_SERVICE);
                    imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
                    listener.onTextChanged(text);
                }
            }
        });

        setOnEditorActionListener(new OnEditorActionListener() {
            @Override
            public boolean onEditorAction(TextView textView, int i, KeyEvent keyEvent) {
                //Keyboard closed with done button
                if (i == EditorInfo.IME_ACTION_DONE) clearFocus();
                return false;
            }
        });

        addTextChangedListener(new SimpleTextWatcher() {
            @Override public void onTextChanged(String newValue) {
                text = newValue;
            }
        });
    }

    @Override
    public boolean onKeyPreIme(int keyCode, KeyEvent event) {
        //Keyboard closed with back button
        if (event.getKeyCode() == KeyEvent.KEYCODE_BACK && event.getAction() == KeyEvent.ACTION_UP) clearFocus();
        return super.dispatchKeyEvent(event);
    }

    public void setTextChangedListener(EditTextListener listener){
        this.listener = listener;
    }
}

File: app/src/main/java/amagi82/modularcharactersheetcreator/extras/widgets/FocusAwareEditText.kt
package amagi82.modularcharactersheetcreator.extras.widgets

import amagi82.modularcharactersheetcreator.extras.widgets.callbacks.EditTextListener
import amagi82.modularcharactersheetcreator.extras.widgets.callbacks.SimpleTextWatcher
import android.content.Context
import android.util.AttributeSet
import android.view.KeyEvent
import android.view.inputmethod.EditorInfo
import android.widget.EditText

class FocusAwareEditText : EditText {
    private var listener: EditTextListener? = null
    private var text: String? = null

    constructor(context: Context) : super(context) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        init()
    }

    private fun init() {
//        onFocusChangeListener = { view, hasFocus ->
//            if (hasFocus)
//                listener.onKeyboardShown()
//            else {
//                //Not focused. Hide keyboard and notify
//                val imm = context.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
//                imm.hideSoftInputFromWindow(view.windowToken, 0)
//                listener.onTextChanged(text)
//            }
//        }

        setOnEditorActionListener { textView, i, keyEvent ->
            //Keyboard closed with done button
            if (i == EditorInfo.IME_ACTION_DONE) clearFocus()
            false
        }

        addTextChangedListener(object : SimpleTextWatcher() {
            override fun onTextChanged(newValue: String) {
                text = newValue
            }
        })
    }

    override fun onKeyPreIme(keyCode: Int, event: KeyEvent): Boolean {
        //Keyboard closed with back button
        if (event.keyCode == KeyEvent.KEYCODE_BACK && event.action == KeyEvent.ACTION_UP) clearFocus()
        return super.dispatchKeyEvent(event)
    }

    fun setTextChangedListener(listener: EditTextListener) {
        this.listener = listener
    }
}